language/evm/move-ethereum-annotator/src/events.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    util::{convert_move_value, ModuleCache},
+    LRU_CACHE_SIZE,
+};
+use anyhow::{anyhow, bail, Result};
+use ethabi::{encode, RawLog};
+use ethereum_types::H256;
+use lru::LruCache;
+use move_core_types::value::MoveValue;
+use move_core_types::{
+    effects::Event, language_storage::TypeTag, resolver::MoveResolver, value::MoveTypeLayout,
+};
+use move_vm_runtime::move_vm::MoveVM;
+
+pub struct MoveEventAnnotator<R: MoveResolver> {
+    resolver: R,
+    abi_cache: ModuleCache<R>,
+    layout_cache: LruCache<TypeTag, MoveTypeLayout>,
+}
+
+impl<R: MoveResolver + Clone> MoveEventAnnotator<R> {
+    pub fn new(resolver: R) -> Self {
+        Self {
+            resolver: resolver.clone(),
+            abi_cache: ModuleCache::new(resolver),
+            layout_cache: LruCache::new(LRU_CACHE_SIZE),
+        }
+    }
+
+    pub fn annotate_move_event(&mut self, event: Event) -> Result<RawLog> {
+        let struct_tag = {
+            use TypeTag::*;
+            match &event.2 {
+                Struct(s) => s,
+                Bool | U8 | U64 | U128 | Address | Signer | Vector(_) => {
+                    bail!("Unsupported primitive types")
+                }
+            }
+        };
+
+        let move_value = {
+            if !self.layout_cache.contains(&event.2) {
+                let move_vm = MoveVM::new(vec![].into_iter()).unwrap();
```

Comment:
Run a VM instance to get MoveTypeLayout may be too heavy?

Reply:
Yea this is a bit unfortunate. I need sth that could convert `StructTag` to `MoveTypeLayout`. I picked MoveVM jsut because it's capabale of doing that. I think this should be fine as the only thing here is the resolver.